### ISSUE TITLE
OData endpoint returns 406 Not Acceptable

### ISSUE THREAD

**Issue by aibars** [Microsoft/aspnet-api-versioning#182]
Hi,

I have successfully run the SwaggerODataWebApiSample example and could hit all of the endpoints.
However, I want to add some OData controllers to my existent versioned Web API solution.

My WebApiConfig.cs looks like this:

 ```
public static void Register(HttpConfiguration config)
        {
            
            config.Formatters.JsonFormatter.SupportedMediaTypes
                .Add(new MediaTypeHeaderValue("text/json"));
            var constraintResolver = new DefaultInlineConstraintResolver()
            {
                ConstraintMap =
                {
                    ["apiVersion"] = typeof( ApiVersionRouteConstraint )
                }
            };
            config.MapHttpAttributeRoutes(constraintResolver);
            config.AddApiVersioning(o => o.ReportApiVersions = true);
            
            GlobalConfiguration.Configuration.Formatters.JsonFormatter.SerializerSettings
                .ReferenceLoopHandling = Newtonsoft.Json.ReferenceLoopHandling.Ignore;

            GlobalConfiguration.Configuration.Formatters
                .Remove(GlobalConfiguration.Configuration.Formatters.XmlFormatter);

            GlobalConfiguration.Configuration.Filters.Add(new CustomExceptionFilterAttribute());

            GlobalConfiguration.Configuration.Filters.Add(new IdentityBasicAuthenticationAttribute());
     }
 ```
And I added the lines of the Startup.cs on your example (excluding the Swagger stuff), like this:

 ```
var modelBuilder = new VersionedODataModelBuilder(config)
            {
                ModelBuilderFactory = () => new ODataConventionModelBuilder().EnableLowerCamelCase(),
                ModelConfigurations =
                {
                    new PersonModelConfiguration()
                }
            };
            var models = modelBuilder.GetEdmModels();

            config.MapVersionedODataRoutes("odata", "api", models, ConfigureODataServices);
            config.MapVersionedODataRoutes("odata-bypath", "api/v{apiVersion}", models, ConfigureODataServices);
```

While compiles OK, I get a 406 Not Acceptable code when I try to access the route:
`api/v1/people(1)` . I have also tried several combinations of the elements of that route without any success.

The route `api/v1/` seems to go to the $metadata controller.

What could I be doing wrong? Is it possible to have a versioned Web API and versioned OData controllers with the same solution configuration?

Thank you.

**Comment by commonsensesoftware** [comment]
First, it's definitely possible to do both. You can look at the [advanced OData Web API sample](../tree/master/samples/webapi/AdvancedODataWebApiSample) for an end-to-end example. One thing to note is that there are limitations in the Web API routing infrastructure. You need to be careful if you use API versioning **and** mix in attribute-based routing for _normal_ controllers. OData only supports convention-based routes under the hood. You _might_ be ok if you don't switch regular and OData controllers for a given service in between API versions. If you switch, then you'll almost certainly need to use convention-based routes. It's kind of lame, but to be sure, you might consider convention-based routes only. I have also found that using that approach helps you troubleshoot routing issues if or when they arise.

A few other minor points:
* In your `Register` method, the **HttpConfiguration** is already provided; prefer the parameter over `GlobalConfiguration.Configuration`
* You do **not** need both flavors of `MapVersionedODataRoutes`. The example shows both merely so I didn't have to create two applications. I suggest that you choose to version by query string or by URL segment, but not both. My recommendation is by query string.
* Why are you registering `text/json`? That is not the official MIME type for JSON; `application/json` is and is already registered. For additional details refer to [this](https://en.wikipedia.org/wiki/JSON#MIME_type) documentation.

Without more information, it's difficult to be sure exactly what's causing the 406, but it is definitely related to media type negotiation. Web API does this by using the registered **MediaTypeFormatter** instances. If I had to guess, you are sending a request to an OData endpoint that has the media type `text/json`, which will **not** work. Furthemore, OData uses its own **MediaTypeFormatter** that does not use or rely on JSON.NET. Since you are receiving 406, this indicates to me that the route selected was for an OData-based controller, but it was unable to format the response. Assuming the routes are configured correctly, sending a request that accepts `application/json` will produce the expected response as is demonstrated in advanced OData example.

I hope that gets you going in the right direction. Let me know if you continue to be stuck.

**Comment by aibars** [comment]
Hi Chris,

First of all, thank so much for your detailed answer. I did take a look at the [advanced OData Web API sample](https://github.com/Microsoft/aspnet-api-versioning/tree/master/samples/webapi/AdvancedODataWebApiSample) and adjusted some things on my `Register` method based on the sample.

I can confirm that the route that is being selected is an OData one (for example, I could hit a breakpoint under the _People_ controller's Get method), but right after that, the response is throwing the 406 error. 
Also, I commented all the `Formatters` related lines with the same results. Regular controllers still work.
I have also tried, comparing all the using references from the controllers and the Startup.cs from the sample and modifying them on my solution. On some cases, I was referencing the `System.Web.Http.OData` namespace instead of `System.Web.Http.OData`. However, after I tried this, I started to get an error at `Global.asax.cs` saying:

`The path template 'People' on the action 'Get' in controller 'People' is not a valid OData path template. Resource not found for the segment 'People'.` 

After that, I noticed that I had some issues in the `PersonModelConfiguration` class. Adjusting those made both OData and Web Api controllers work.

Regarding the points you mentioned:

1) Yes, you're right. I guess I got distracted and forgot to modify that. 
2) I tried one of the `MapVersionedODataRoutes` forms but had the same results.
3) Don't know exactly why was this configured like that (could be some front-end related reason).

It seems that with your guidance combined to checking the sample solutions enabled me to solve the issue.

Thank you.
Agustín
